Murloc11/bjs2 83995

bookinisto opened the pull request.

bookinisto committed:
сделал :
-Создание скилла
-Получение списка скиллов пользователя
-Просмотр предложенных скиллов
осталось :
-Приобретение предложенных скиллов

bookinisto committed:
сделал :
-Создание скилла
-Получение списка скиллов пользователя
-Просмотр предложенных скиллов
-Приобретение предложенных скиллов

bookinisto committed:
сделал :
-Создание скилла
-Получение списка скиллов пользователя
-Просмотр предложенных скиллов
-Приобретение предложенных скиллов

bookinisto committed:
сделал :
-Создание скилла
-Получение списка скиллов пользователя
-Просмотр предложенных скиллов
-Приобретение предложенных скиллов

bookinisto committed:
сделал :
-Создание скилла
-Получение списка скиллов пользователя
-Просмотр предложенных скиллов
-Приобретение предложенных скиллов

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+@Component
```
Смотри, да это компонент. Но это именно контроллер, тебе нужно заменить Аннотацию компонент на @RestController и  добавить @RequestMapping("/skills") - это аннотация самого эндпоинта (Адреса, по которому ты будешь вызывать методы)

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,41 @@
+    private final SkillService skillService;
+    private final UserContext userContext;
+
+    public SkillDto create(CreateSkillDto skillDto) {
```
Это методы CRUD операций, почитай об этом подробнее.
https://www.youtube.com/watch?v=vD0X5Zm9Gjo 
Вот неплохое видео. Суть в том, что тебе нужно проставить тут Аннотации Spring на методах - Получения, обновления, Изменения и удаления.

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,41 @@
+        return skillService.create(skillDto);
+    }
+
+    private void validateSkillNameNotNull(String title) {
```
Метод валидации не должен быть в контроллере, это нарушает принцип единой ответственности, методы валидации нужно вынести в отдельный класс или обрабатывать их через GlobalExceptionHandler (Об этом почитай, будет дальше в курсе)

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    public List<SkillDto> getByUserId(Long userId) {
```
Методы контроллера должны возвращать ResponseEntity<T>
Вот пример: 

```@PostMapping
    public ResponseEntity<EventViewDto> create(@Valid @RequestBody EventCreateDto eventDto) {
        return ResponseEntity.ok(eventService.create(eventDto));
    }

JekaCAP commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public interface SkillService {
```
java doc нет, да он не обязательный, но очень помогает читать твой код. 
Вот отличная статья на эту тему: https://java-online.ru/java-javadoc.xhtml

JekaCAP reviewed (commented):
Привет! В целом код написал хорошо, нужно исправить некоторые моменты, если есть вопросы, пиши в лс, созвонимся, помогу разобраться :) 
оставлю ребятам, ещё часть кода твоего на проверку.

bookinisto committed:
сделал :
-валидацию на null и пустое имя в SkillDto аннотацией  @NotBlank
-@RequestMapping с GET POST PUT запросами в SkillController

bookinisto committed:
BJS2-83995
сделал :
-для SkillServiceImpl и SkillController создал классы SkillServiceValidator и SkillControllerValidator с методами валидации
и реализовал их в классах сервиса и контроллера

bookinisto committed:
BJS2-83995
сделал :
-Обёртывание ответа методов SkillController  в ResponseEntity

bookinisto committed:
BJS2-83995
сделал :
-Обёртывание ответа методов SkillController  в ResponseEntity

bookinisto committed:
BJS2-83995
сделал :
-Обёртывание ответа методов SkillController  в ResponseEntity

bookinisto committed:
BJS2-83995
сделал :
-Обёртывание ответа методов SkillController  в ResponseEntity

bookinisto committed:
BJS2-83995
сделал :
-заменил в методах передачу в параметре примитивные типы на объекты обёртки
-сократил название полей класов

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+                : ResponseEntity.ok(skills);
+    }
+
+    @GetMapping("/offered")
```
Аннотации эндпоинтов не правильные, у тебя основная @RequestMapping("/skills")
А добавляется только там, где методы по id скилла вызываются

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+                .body(createdSkill);
+    }
+
+    @GetMapping("/user/{userId}")
```
Тут ты получается обращаешься к пользователю, хотя действия со скиллами

Вот тут например просто @PutMapping("/{userId")

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+                : ResponseEntity.ok(skills);
+    }
+
+    @PutMapping("/acquire")
```
А тут @PutMapping("/{skillId")

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillControllerValidator.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.stereotype.Controller;
+import school.faang.user_service.exception.DataValidationException;
+
+@Controller
```
Тут аннотация спринга не нужна, ты не внедряешь этот класс как бин не где

JekaCAP commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,10 @@
+public record SkillDto(
+        Long id,
+        @NotBlank(message = "Имя скилла не может быть пустым")
+        String title
```
Валидация имеет смысл только в DTO обновления и добавления

JekaCAP commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface SkillService {
+
+    SkillDto create(CreateSkillDto skillDto);
```
Файл как делать java doc скинул в чат нашей группы в DS, если не сделаешь её, я всё равно аппрувну

asQ32 commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+public class SkillController {
+    private final SkillService service;
+    private final UserContext context;
+    private final SkillControllerValidator controllerValidator;
```
controllerValidator -> validator

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @PutMapping("/acquire")
+    public ResponseEntity<Void> acquireSkillFromOffers(@RequestParam Long skillId) {
```
На id используй в параметрах @PathVariable

JekaCAP commented on src/main/java/school/faang/user_service/service/skill/SkillServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+public class SkillServiceImpl implements SkillService {
+
+    @Value("${skill.count-recommendation.min-length}")
+    private static int countOfSkillRecommendation;
```
Убери статик иначе не будет работать

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillControllerValidator.java:
```diff
@@ -0,0 +1,23 @@
+            throw new DataValidationException("'%s' должен быть числом".formatted(object));
+        }
+
+        long id = ((Number) object).longValue();
```
Для простой валидации данных можно использовать Jakarta Validation

JekaCAP commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @GetMapping("/user/{userId}")
+    public ResponseEntity<List<SkillDto>> getByUserId(@PathVariable Long userId) {
```
id на null - уже проверяет сам валидатор

JekaCAP reviewed (commented):
Отличные правки!!! осталось совсем чуть чуть, ты можешь написать мне в ТГ и вместе сделаем правки, я помогу и объясню. Если у тебя Discord сильно тупит 
@EvgeniyStaff - мой телеграмм, пиши :)

bookinisto committed:
BJS2-83995
сделал :
- добавил @RequiredArgsConstructor в class SkillServiceValidator
- скорректировал http запросы в контроллере
- убрал лишний класс валидации конроллера

bookinisto committed:
BJS2-83995
сделал :
- добавил @RequiredArgsConstructor в class SkillServiceValidator
- скорректировал http запросы в контроллере
- убрал лишний класс валидации конроллера

RinatBeybutov commented on gradle/wrapper/A.java:
```diff
@@ -0,0 +1,3 @@
+PS C:\Users\booki\IdeaProjects\user_service> java -jar app.jar
+Error: Unable to access jarfile app.jar
```
а это для чего?

bookinisto committed:
BJS2-83995
сделал :
- удалил странный jar файл
